compiler/src/main.rs

```diff
@@ -0,0 +1,208 @@
+    let ptx = str::from_utf8(&ptx).map_err(CompilerError::from)?;
+    let llvm = ptx_to_llvm(ptx).map_err(CompilerError::from)?;
+
+    let output = match output_type {
+        OutputType::LlvmIr => {
+            if opts.linked {
+                get_linked_bitcode(&llvm)?
+            } else {
+                llvm.llvm_ir
+            }
+        }
+        OutputType::Elf => get_elf(&llvm)?,
+        OutputType::Assembly => get_assembly(&llvm)?,
+    };
+
+    write_to_file(&output, &output_path).map_err(CompilerError::from)?;
+    Ok(())
+}
+
+fn ptx_to_llvm(ptx: &str) -> Result<LLVMArtifacts, CompilerError> {
+    let ast = ptx_parser::parse_module_checked(ptx).map_err(CompilerError::from)?;
+    let module = ptx::to_llvm_module(ast).map_err(CompilerError::from)?;
+    let bitcode = module.llvm_ir.write_bitcode_to_memory().to_vec();
+    let linked_bitcode = module.linked_bitcode().to_vec();
+    let llvm_ir = module.llvm_ir.print_module_to_string().to_bytes().to_vec();
+    Ok(LLVMArtifacts {
+        bitcode,
+        linked_bitcode,
+        llvm_ir,
+    })
+}
+
+#[derive(Debug)]
+struct LLVMArtifacts {
+    bitcode: Vec<u8>,
+    linked_bitcode: Vec<u8>,
+    llvm_ir: Vec<u8>,
+}
+
+fn get_arch() -> Result<CString, CompilerError> {
+    use hip_runtime_sys::*;
+    unsafe { hipInit(0) }?;
+    let mut dev_props: MaybeUninit<hipDeviceProp_tR0600> = MaybeUninit::uninit();
+    unsafe { hipGetDevicePropertiesR0600(dev_props.as_mut_ptr(), 0) }?;
+    let dev_props = unsafe { dev_props.assume_init() };
+    let arch = dev_props.gcnArchName;
+    let arch: Vec<u8> = arch
+        .to_vec()
+        .iter()
+        .map(|&v| i8::to_ne_bytes(v)[0])
+        .collect();
+    let arch = CStr::from_bytes_until_nul(arch.as_slice())?;
+    Ok(CString::from(arch))
+}
+
+fn get_linked_bitcode(llvm: &LLVMArtifacts) -> Result<Vec<u8>, CompilerError> {
+    let linked_bitcode = comgr::get_linked_bitcode_as_bytes(&llvm.bitcode, &llvm.linked_bitcode)?;
+    Ok(ptx::bitcode_to_ir(linked_bitcode))
+}
+
+fn get_elf(llvm: &LLVMArtifacts) -> Result<Vec<u8>, CompilerError> {
+    let arch = get_arch()?;
+    comgr::get_executable_as_bytes(&arch, &llvm.bitcode, &llvm.linked_bitcode)
+        .map_err(CompilerError::from)
+}
+
+fn get_assembly(llvm: &LLVMArtifacts) -> Result<Vec<u8>, CompilerError> {
+    let arch = get_arch()?;
+    comgr::get_assembly_as_bytes(&arch, &llvm.bitcode, &llvm.linked_bitcode)
+        .map_err(CompilerError::from)
+}
+
+fn check_path(path: &Path) -> Result<(), CompilerError> {
+    if path.try_exists().map_err(CompilerError::from)? && !path.is_file() {
+        let message = format!("Not a regular file: {:?}", path.to_path_buf());
+        let error = CompilerError::GenericError {
+            cause: None,
+            message,
+        };
+        return Err(error);
+    }
+    Ok(())
+}
+
+fn get_output_path(ptx_path: &PathBuf, output_type: &OutputType) -> Result<PathBuf, CompilerError> {
```

Comment:
By default we should build in the folder where .ptx is located. That's because the default use is you are going to do multiple runs of `zluda_dump` which will write out all teh PTX to separate directories. So you will have multiple ptx from different runs or and different applications: `C:\temp\zluda\geekbench_x86_64.exe\module_0001_05.ptx`, `C:\temp\zluda\geekbench_x86_64.exe_1\module_0001_05.ptx`, `C:\temp\zluda\python.exe\module_0001_05.ptx`.
It makes it easier to then navigate if the zoc output is then right next to ptx